src/mlpack/core/optimizers/katyusha/katyusha_impl.hpp

```diff
@@ -0,0 +1,202 @@
+  arma::mat fullGradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+
+  arma::mat iterate0 = iterate;
+  arma::mat y = iterate;
+  arma::mat z = iterate;
+  arma::mat w = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate0, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Katyusha: converged to " << overallObjective
+          << "; terminating  with failure.  Try a smaller step size?"
+          << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "Katyusha: minimized within tolerance " << tolerance
+          << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
+    for (size_t f = effectiveBatchSize; f < numFunctions;
+        /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate0, f, gradient, effectiveBatchSize);
+      fullGradient += gradient;
+
+      f += effectiveBatchSize;
+    }
+    fullGradient /= (double) numFunctions;
+
+    // To keep track of where we are and how things are going.
+    double cw = 1;
+    w.zeros();
+
+    for (size_t f = 0, currentFunction = 0; f < innerIterations;
+        /* incrementing done manually */)
+    {
+      // Is this iteration the start of a sequence?
+      if ((currentFunction % numFunctions) == 0)
+      {
+        currentFunction = 0;
+
+        // Determine order of visitation.
+        if (shuffle)
+          function.Shuffle();
+      }
+
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - currentFunction);
+      iterate = tau1 * z + tau2 * iterate0 + (1 - tau1 - tau2) * y;
+
+      // Calculate variance reduced gradient.
+      function.Gradient(iterate, currentFunction, gradient,
+          effectiveBatchSize);
+      function.Gradient(iterate0, currentFunction, gradient0,
+          effectiveBatchSize);
+
+      arma::mat zNew = z - alpha * (fullGradient + (gradient - gradient0) /
+          (double) batchSize);
```

Comment:
I believe this comes from the top of page 11:

> By the minimality definition of z_{k + 1}, we have that
>   z_{k+1} − z_k + α∇_{k+1} + αg = 0

Do you think it would be useful to add a comment to point that out?  I spent a little while looking at the original line 10 in the Katyusha() algorithm before I realized the equivalence here.